pkg/sink/sink.go

```diff
@@ -0,0 +1,130 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sink
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"path"
+	"strings"
+
+	triggersclientset "github.com/tektoncd/triggers/pkg/client/clientset/versioned"
+	"github.com/tektoncd/triggers/pkg/template"
+	"github.com/tidwall/gjson"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeclientset "k8s.io/client-go/kubernetes"
+)
+
+type Resource struct {
+	TriggersClient         *triggersclientset.Clientset
+	KubeClient             *kubeclientset.Clientset
+	EventListenerName      string
+	EventListenerNamespace string
+}
+
+func (r Resource) HandleEvent(response http.ResponseWriter, request *http.Request) {
+	el, err := r.TriggersClient.TektonV1alpha1().EventListeners(r.EventListenerNamespace).Get(r.EventListenerName, metav1.GetOptions{})
+	if err != nil {
+		log.Printf("Error getting EventListener %s in Namespace %s: %s", r.EventListenerName, r.EventListenerNamespace, err)
+		return
+	}
+
+	event, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		log.Printf("Error reading event body: %s", err)
+		return
+	}
+
+	for _, tbRef := range el.Spec.TriggerBindingRefs {
+		tbNamespace := tbRef.Namespace
+		if tbNamespace == "" {
+			tbNamespace = r.EventListenerNamespace
+		}
+		tb, err := r.TriggersClient.TektonV1alpha1().TriggerBindings(tbNamespace).Get(tbRef.Name, metav1.GetOptions{})
+		if err != nil {
+			log.Printf("Error getting TriggerBinding %s in Namespace %s: %s", tbRef.Name, tbNamespace, err)
+			continue
+		}
+
+		for _, templateBinding := range tb.Spec.TemplateBindings {
+			ttNamespace := templateBinding.TemplateRef.Namespace
+			if ttNamespace == "" {
+				ttNamespace = r.EventListenerNamespace
+			}
+			tt, err := r.TriggersClient.TektonV1alpha1().TriggerTemplates(ttNamespace).Get(templateBinding.TemplateRef.Name, metav1.GetOptions{})
+			if err != nil {
+				log.Printf("Error getting TriggerTemplate %s in Namespace %s: %s", templateBinding.TemplateRef.Name, ttNamespace, err)
+				continue
+			}
+
+			params, err := template.ApplyEventToParams(event, templateBinding.Params)
+			if err != nil {
+				log.Printf("Error applying event to TriggerBinding params: %s", err)
+				continue
+			}
+			params = template.AddDefaultParamsFromSpecs(params, tt.Spec.Params)
+
+			for _, rt := range tt.Spec.ResourceTemplates {
+				rtRaw := template.ApplyParamsToResourceTemplate(params, rt.RawMessage)
+				err = createResource(r.KubeClient, rtRaw)
+				if err != nil {
+					log.Printf("Error creating resources from resource template %s: %s", string(rtRaw), err)
+					continue
+				}
+				log.Printf("Successfully created resource: %s", string(rtRaw))
+			}
+		}
+	}
+}
```

Comment:
`HandleEvent` looks like a great candidate to break up into separate functions with unit tests - at the moment it would probably be difficult to write unit tests for `HandleEvent` will full coverage (I think we should probably have at least one tho!)

One thing that we do in Pipelines that (mostly) works is separating following all the references in a CRD from validating + from taking any action.

So it could be interesting to structure this function as something like (where each line is a separate function)

- Get all the CRD instances (i.e. follow all the Refs) needed -> probably return a new object that contains all the referenced objects together, e.g. slice of triggerbinding + triggertemplate
- Apply all the events (i.e. do the interpolation, `ApplyEventToParams`, `AddDefaultParamsFromSpecs`, `ApplyParamsToResourceTemplate`
- Create all the Resources

Ideally `HandleFunc` could tell us at a glance what all is being done each time its called, e.g. here's one idea of how it could look:
```go

type ResolvedBinding struct {
    TriggerBinding b
    TriggerTemplate t
}

func (r Resource) HandleEvent(response http.ResponseWriter, request *http.Request) {
	el, err := r.TriggersClient.TektonV1alpha1().EventListeners(r.EventListenerNamespace).Get(r.EventListenerName, metav1.GetOptions{})
	if err != nil {
		log.Printf("Error getting EventListener %s in Namespace %s: %s", r.EventListenerName, r.EventListenerNamespace, err)
		return
	}

 	event, err := ioutil.ReadAll(request.Body)
	if err != nil {
		log.Printf("Error reading event body: %s", err)
		return
	}

        // Slice of ResolvedBinding
        bindings, err := ResolveBindings(el.Spec.TriggerBindingRefs, r.TriggersClient.TektonV1alpha1().TriggerBindings(tbNamespace).Get,
r.TriggersClient.TektonV1alpha1().TriggerTemplates(ttNamespace).Get)
      if err {}

       // Some object that contains the resources to create, maybe slice of json.RawMessage?
       resourcesToCreate, err := GetNewResources(bindings)
       if err {}

       if err := CreateResources(resourcesToCreate, kubeClient); err != nil { ... }<-- kubeClient interface looks v. hard to mock :S

```

Also note that a handy trick for functions that need to use the clients is to pass the functions around, e.g.

https://github.com/tektoncd/pipeline/blob/b8100856fedda3cb77f31b9487d093af5597fc7c/pkg/reconciler/taskrun/resources/taskresourceresolution.go#L38-L44

This lets unit tests create their own `GetResource` functions which is way less hassle than mocking clients.

So we'd be able to make unit tests (hopefully!) fairly easily for `ResolveBindings` and `GetNewResources` (and probably `CreateResources` too but I dunno how to mock out that RESTClient :thinking: )